### ISSUE TITLE
APNS: Errno 104: Connection reset by peer

### ISSUE THREAD

**Issue by Lukas0907** [dgilland/pushjack#19]
Hi,

I want to send 1.500 push notifications via APNS with the default batch size of 100 but the server seems to close the connection at some point:

```
Sending APNS notification batch containing 12900 bytes.
Traceback (most recent call last):
  File "pushjack-test.py", line 20, in <module>
    res = client.send(tokens, alert)
  File "redacted/pushjack/pushjack/apns.py", line 241, in send
    return self.conn.sendall(stream, error_timeout)
  File "redacted/pushjack/pushjack/apns.py", line 411, in sendall
    self.send(stream)
  File "redacted/pushjack/pushjack/apns.py", line 396, in send
    self.write(frame)
  File "redacted/pushjack/pushjack/apns.py", line 362, in write
    return self.client.sendall(data)
  File "/usr/lib/python2.7/ssl.py", line 329, in sendall
    v = self.send(data[count:])
  File "/usr/lib/python2.7/ssl.py", line 298, in send
    v = self._sslobj.write(data)
socket.error: [Errno 104] Connection reset by peer
```

It doesn't always happen but most of the time, when I try to send more than about 1000 messages. It always works for less than that.

I've implemented some cheap reconnection mechanism which seems to fix that problem:

``` diff
diff --git a/pushjack/apns.py b/pushjack/apns.py
index dda4f51..d4cc98a 100644
--- a/pushjack/apns.py
+++ b/pushjack/apns.py
@@ -393,7 +396,14 @@ class APNSConnection(object):
     def send(self, frames):
         """Send stream of frames to APNS server."""
         for frame in frames:
-            self.write(frame)
+            while True:
+                try:
+                    self.write(frame)
+                    break
+                except socket.error as ex:
+                    log.error(('Could not send frame to server: {}'.format(ex)))
+                    self.close()
+                    self.connect()
             self.check_error(0)

     def sendall(self, stream, error_timeout=10):
```

From the log output:

```
[...]
Preparing notification for APNS token xxx
Sending APNS notification batch containing 12900 bytes.
Could not send frame to server: [Errno 104] Connection reset by peer
Closing connection to APNS.
Establishing connection to APNS on gateway.push.apple.com:2195 using certificate at mc-apns.key.pem
Performing SSL handshake with APNS on gateway.push.apple.com:2195
Established connection to APNS on gateway.push.apple.com:2195.
Sending APNS notification batch containing 12900 bytes.
Preparing notification for APNS token xxx
Preparing notification for APNS token xxx
[...]
```

Regards,

Lukas

**Comment by dgilland** [comment]
Do you know if any of the device tokens you were sending were invalid, e.g., a mixture sandbox vs production tokens? That's sometimes why APNS will disconnect during sending.

The only thing I might change with your implementation would be to add a retry limit to avoid an infinite disconnect/reconnect loop in case something really goes wrong.

I've tried to avoid putting retry logic into pushjack since I wanted to keep the implementation simpler but admittedly there is not a good way from the outside to resend failed tokens. See #10. Obviously, at a minimum pushjack should provide a way for the user to resume sending from the last failure point but I haven't had time to implement it properly.

**Comment by jakubczaplicki** [comment]
This could also happen when some of the tokens you're trying to send have expired.

**Comment by Lukas0907** [comment]
> Do you know if any of the device tokens you were sending were invalid, e.g., a mixture sandbox vs production tokens? That's sometimes why APNS will disconnect during sending.

The problem originally happened in production where all tokens are real world production tokens. Some of them could have been expired if the user has uninstalled the app before we tried to send them a notification.

For my tests I used the production server and about 1500 times one token, i.e. the iPhone receives this many push notifications as well. Since this causes the same symptoms, the problem might not be related to the token. Anyway it would be nice if socket errors are handled gracefully since they can also happen due to (network) outages (but I'm aware that this is the scope of #10).
